quinn-proto/src/connection/spaces.rs

```diff
@@ -129,12 +130,12 @@ impl PacketSpace {
             self.immediate_ack_pending = true;
         }
 
-        // Retransmit the data of the oldest in-flight packet
-        if !self.pending.is_empty(streams) {
+        if !self.pending.is_empty(streams) || !datagrams.outgoing.is_empty() {
```

Comment:
I don't think this is sufficient, because the next user datagram might not be able to fit in the next QUIC packet.  And in that case you'd still have to include a ping frame.  I guess the bug in this scenario is worse than sending the extra ping frame.

I also think it would be really good to write a test for this.

Reply:
Hmm, good catch. I think we actually want exactly the same judgement as `Connection::can_send_1rtt` here. It looks like that would actually fix another pre-existing bug here, since lazy cleanup of reset streams means `self.pending.is_empty` doesn't actually necessarily imply the ability to transmit any stream frames.

Reply:
TIL, I assumed the stream side of this was correct.

Reply:
I'm actually waffling a bit on whether the datagram side is an issue or not. The interaction of TLP queue accounting and size limits is nontrivial. I think I'll need to build some tests to draw empirical conclusions and validate a solution.

Reply:
I'm wondering whether the best thing is to revert this change and create an issue for the extra ping.  That way we have a correct-but-slightly-wasteful implementation and don't forget to follow up.

Reply:
I'm wondering whether the best thing is to revert this change and create an issue for the extra ping.  That way we have a correct-but-slightly-wasteful implementation and don't forget to follow up.